packages/labs/analyzer/src/lib/program-context.ts

```diff
@@ -0,0 +1,478 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import ts from 'typescript';
+import {DiagnosticsError, createDiagnostic} from './errors.js';
+import * as typedoc from 'typedoc';
+import path from 'path';
+import fs from 'fs';
+import {Module, PackageJson, Type, Reference} from './model.js';
+
+type FileCache = ts.MapLike<{version: 0; content?: ts.IScriptSnapshot}>;
+
+/**
+ * Create a language service host that reads from the filesystem initially,
+ * and supports updating individual source files in memory by updating its
+ * content and version in the FileCache.
+ */
+const createServiceHost = (
+  commandLine: ts.ParsedCommandLine,
+  cache: FileCache
+): ts.LanguageServiceHost => {
+  return {
+    getScriptFileNames: () => commandLine.fileNames,
+    getScriptVersion: (fileName) => cache[fileName]?.version.toString(),
+    getScriptSnapshot: (fileName) => {
+      let file = cache[fileName];
+      if (!file) {
+        if (!fs.existsSync(fileName)) {
+          return undefined;
+        }
+        file = {
+          version: 0,
+          content: ts.ScriptSnapshot.fromString(
+            fs.readFileSync(fileName).toString()
```

Comment:
if you give an encoding the return value is a String already, I think requiring `utf8` is fine.

Reply:
+1 Done